backupccl: add SHOW BACKUP option for validating descriptors

fqazi opened the pull request:
Fixes: #85268

Previously, the only way to know the validity of a backup
was to restore the entire backup. This was inadequate because
there was no way to know if a given backup would restore without
any descriptor validation errors. To address this, this patch
adds support for validating the descriptors inside a given
backup as a hidden option.

This validation like the existing debug doctor command will
not be perfect, and is subject to interpretation.

Release justification: low risk command for validating descriptors
inside a backup. Only for internal use and not a user facing
feature.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86817)
<!-- Reviewable:end -->

fqazi committed:
backupccl: add SHOW BACKUP option for validating descriptors

Fixes: #85268

Previously, the only way to know the validity of a backup
was to restore the entire backup. This was inadequate because
there was no way to know if a given backup would restore without
any descriptor validation errors. To address this, this patch
adds support for validating the descriptors inside a given
backup as a hidden option.

This validation like the existing debug doctor command will
not be perfect, and is subject to interpretation.

Release justification: low risk command for validating descriptors
inside a backup. Only for internal use and not a user facing
feature.

Release note: None

msbutler reviewed (commented):
Looks good from the bulk perspective! I did not look at the innards of backupShowDoctor's function or the changes to doctor.go -- that should get reviewed by a schema person. 

Given this feature is internal only, I support the proposed syntax of 
SHOW BACKUP VALIDATE

If this feature ever became generally available, we'd have to think a bit harder about the exact syntax of this feature.

Xiang-Gu commented on pkg/ccl/backupccl/testdata/show_backup_validate/show_backup:
```diff
@@ -0,0 +1,15 @@
+validate-backup invalidDependOn_21.1
```
the error message seems to suggest it's a  `invalidDependedOnBy` case, so should we rename it to `invalidDependedOnBy_21.1`?

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @Xiang-Gu)

___
*[`pkg/ccl/backupccl/show_test.go` line 1027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzkg-r1-1027:-NAKrqmkAMDVwrIvA-ZJ:b-r87d7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/ccl/backupccl/show_test.go#L1027)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

this test should use the data driven driver that other data driven backup tests use. Before that happens, some test infra refactoring that needs to occur, and I don't think you need to do that. After the refactor, there shouldn't be a need for a special "validate-backup" cmd. The test file will only only need the actual sql.

Could you add a "TODO(msbutler): incorporate this test into backup data driven driver"
</blockquote></details>

I moved this completely into the datadrive portion and just added a command to link the file. Think that's cleaner and more maintainable in the long term.
___
*[`pkg/ccl/backupccl/testdata/show_backup_validate/show_backup` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzkn-r1-1:-NAKrwa_6EscltauQRIf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/ccl/backupccl/testdata/show_backup_validate/show_backup#L1)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

the error message seems to suggest it's a  `invalidDependedOnBy` case, so should we rename it to `invalidDependedOnBy_21.1`?
</blockquote></details>

Done.
___
*[`pkg/sql/parser/sql.y` line 6625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzks-r1-6625:-NAKTgx09qTTHPUuhBsO:b-36hut3) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/sql/parser/sql.y#L6625)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Does this feature need to support old style backups (old: BACKUP TO vs new: BACKUP INTO)? If so, you may want to test for it, at least by running a 1 off test on your own. If not, feel free to delete this parsing option. 

You may want to support the old style as backupper still uses it. 
</blockquote></details>

Let me add test coverage for the old syntax


<!-- Sent from Reviewable.io -->

msbutler reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @Xiang-Gu)

___
*[`pkg/ccl/backupccl/datadriven_test.go` line 371 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAPcm8D6-NUELStBkDQ:-NAPcm8D6-NUELStBkDR:bqk7sj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/272236510a771a8352452aa607fd72be6d4d2ad7/pkg/ccl/backupccl/datadriven_test.go#L371)):*
> ```Go
> //   - "corrupt-backup" uri=<collectionUri>
> //     Finds the latest backup in the provided collection uri an flips a bit in one SST in the backup
> //
> ```

this is great!
___
*[`pkg/ccl/backupccl/show_test.go` line 1027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzkg-r1-1027:-NAPbOgb6m1sB-Sp3Se4:b9bkkp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/ccl/backupccl/show_test.go#L1027)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I moved this completely into the datadrive portion and just added a command to link the file. Think that's cleaner and more maintainable in the long term.
</blockquote></details>

Looks great!
___
*[`pkg/ccl/backupccl/testdata/backup-restore/show_backup` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAPbt4D1-W5yugWZcse:-NAPbt4D1-W5yugWZcsf:b-ijuiib) ([raw file](https://github.com/cockroachdb/cockroach/blob/272236510a771a8352452aa607fd72be6d4d2ad7/pkg/ccl/backupccl/testdata/backup-restore/show_backup#L3)):*
> ```
> new-server name=s1 allow-implicit-access
> ----
> ```

Could you add a comment up here that explains the basic design of the test? i.e. "this test check that SHOW BACKUP VALIDATE properly detects invalid descriptors using the new and old SHOW BACKUP syntax"
___
*[`pkg/sql/parser/sql.y` line 6625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzks-r1-6625:-NAPcKoBA3smNS9FSqGo:b-vs1prh) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/sql/parser/sql.y#L6625)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Let me add test coverage for the old syntax
</blockquote></details>

Looks good!


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
New changes look great. Will wait for a schema engineer to approve this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-:-NAPdMPDDCqE8b2Vedad:b-52qjcz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

fqazi commented:
@msbutler TFTR

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @msbutler, and @Xiang-Gu)

___
*[`pkg/ccl/backupccl/datadriven_test.go` line 371 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAPcm8D6-NUELStBkDQ:-NAQHZ6u2VzNr2Gbhchy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/272236510a771a8352452aa607fd72be6d4d2ad7/pkg/ccl/backupccl/datadriven_test.go#L371)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

this is great!
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/show_test.go` line 1027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzkg-r1-1027:-NAQHZYH9r1MQuIZLgwe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/ccl/backupccl/show_test.go#L1027)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Looks great!
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/testdata/backup-restore/show_backup` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAPbt4D1-W5yugWZcse:-NAQH_CJ4C6EHGnM_MTA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/272236510a771a8352452aa607fd72be6d4d2ad7/pkg/ccl/backupccl/testdata/backup-restore/show_backup#L3)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Could you add a comment up here that explains the basic design of the test? i.e. "this test check that SHOW BACKUP VALIDATE properly detects invalid descriptors using the new and old SHOW BACKUP syntax"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (approved):
LGTM with one nit!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-:-NAdt53IDWaoVYBv8rj9:b-jz2jqj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @msbutler, and @Xiang-Gu)

___
*[`pkg/cli/doctor.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAdsyFO0dZqG-91WmQs:-NAdsyFO0dZqG-91WmQt:be0igej) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc688dd0384bad04f995c702050272a3c2679c01/pkg/cli/doctor.go#L179)):*
> ```Go
> 		namespaceTable,
> 		jobsTable,
> 		true,
> ```

nit: add a /* validateJobs */ after this `true` literal


<!-- Sent from Reviewable.io -->

fqazi commented:
@Xiang-Gu TFTR!

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86817)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @Xiang-Gu)

___
*[`pkg/sql/parser/sql.y` line 6625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86817#-NAGgMug0IFZpSD2Bzks-r1-6625:-NAe3eKdAgdN6w9qICOO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/639bad55a57a2a39eab713e95b1de6d935177b88/pkg/sql/parser/sql.y#L6625)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Looks good!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
